src/test/scala/com/advancedtelematic/campaigner/db/CampaignsSpec.scala

```diff
@@ -75,40 +74,34 @@ final class CampaignsFindFailedDevicesSpec extends AsyncFlatSpec
   with DatabaseUpdateSpecUtil
   with CampaignerSpecUtil {
 
-  import Arbitrary._
-
   val campaigns = Campaigns()
-  implicit val defaultPatience = PatienceConfig(timeout = Span(2, Seconds))
+  implicit val defaultPatience = PatienceConfig(timeout = 2 seconds)
 
   "findFailedDeviceUpdates" should "find all failed device update from all the given campaigns and return only the most recent ones" in {
-    val nCampaigns = Gen.choose(1, 4).generate
-    val campaignIds = Gen.listOfN(nCampaigns, genCampaignId).generate
-    val resultCodes = Seq("RC1", "RC2", "RC3")
 
+    val campaignIds = Gen.choose(3, 6).flatMap(Gen.listOfN(_, genCampaignId)).generate
     val updateId = createDbUpdate(UpdateId.generate()).futureValue
-    val campaignObjects = campaignIds.map(cid =>
-        arbitrary[Campaign].generate.copy(id = cid, updateId = updateId))
+    val campaignObjects = campaignIds.map(cid => genCampaign.generate.copy(id = cid, updateId = updateId))
 
-    val nUpdates = Gen.choose(10, 20).generate
-    val updates = Gen.listOfN(nUpdates, genDeviceUpdate(
-      genCampaignId = Gen.oneOf(campaignIds),
-      genResultCode = Gen.oneOf(resultCodes),
-    )).generate
+    val deviceIds = Gen.choose(2, 8).flatMap(Gen.listOfN(_, genDeviceId)).generate
+    val updates = campaignIds.flatMap(cid => deviceIds.map { did =>
+      genDeviceUpdate().map(_.copy(campaign = cid, update = updateId, device = did)).generate
+    })
 
-    val byUpdatedAtDesc = Ordering.by[DeviceUpdate, Long](_.updatedAt.getEpochSecond)
     val expectedFailures = updates
+      .filter(du => du.status != DeviceStatus.requested && du.status != DeviceStatus.scheduled && du.status != DeviceStatus.accepted)
       .groupBy(_.device)
-      .mapValues(_.sorted(ord = byUpdatedAtDesc).head)
+      .mapValues(_.maxBy(_.updatedAt))
       .values
       .filter(_.status == DeviceStatus.failed)
       .toSet
 
-    db.run(for {
+    val resultAction = for {
       _ <- Schema.campaigns ++= campaignObjects
       _ <- Schema.deviceUpdates ++= updates
       failures <- campaigns.findFailedDeviceUpdatesAction(campaignIds.toSet)
-    } yield failures).flatMap { actualFailures =>
-      actualFailures should contain theSameElementsAs expectedFailures
-    }
+    } yield failures
+
+    db.run(resultAction).flatMap(_ should contain theSameElementsAs expectedFailures)
```

Comment:
_Side note_: could you please create a ticket and put a TODO to refactor all the `AsyncFlatSpec` tests? They are quite tedious to work with, IMO, and I don't see any advantages in using them instead of `ScalaFutures` mixin. If you know why `AsyncFlatSpec` might be a good deal, I'd be glad to hear it.

Reply:
AFIAU the main (only?) advantage is that we don't need to worry about tests failing because of a timeout when waiting for a Future to complete. I've seen a few of those and they're annoying (and the reason we need to increase the patience sometimes).
On the other hand, they _are_ tedious and make it more difficult to use property-based testing. So I guess we need to pick our poison.

Reply:
I think the timeout from PatienceConfig is also applied when using `futureValue` method of `ScalaFutures` trait, so it's possible to tweak it too. At least that's the impression I get from reading http://doc.scalatest.org/3.0.1/#org.scalatest.concurrent.ScalaFutures

Reply:
Indeed, but sometimes it's a bit tricky to find the right value for the patience without setting it to something ludicrous 😕 For instance in this test, I tried to use `futureValue` and it was timing out most of the times with a generous patience. 
Anyways, I'm also in favour of `ScalaFutures`, it's easier to use and to follow (and to use property-based tests). Just saying that maybe once in a while `AsyncFlatSpec` is worth the trouble.